<a href="https://colab.research.google.com/github/luciazarpe/TFM/blob/main/eda.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **TFM DSMarket — Análisis Exploratorio de Datos (EDA)**

Notebook de análisis exploratorio del dataset DSMarket. Examinamos la evolución temporal de las ventas, el comportamiento por tienda, el efecto de los días festivos y eventos, y el rendimiento de los productos, con el objetivo de extraer conclusiones que guíen la modelización.

---

## **Tabla de contenidos**

1. [Análisis temporal](#1-análisis-temporal)
   - Evolución de ingresos diarios
   - Medias móviles
   - Ingresos por trimestre
   - Efecto fin de semana
   - Heatmap día × mes
   - Efecto período de cobro

2. [Análisis por tienda](#2-análisis-por-tienda)
   - Tendencia de ingresos
   - Patrón semanal
   - Distribución de ventas diarias
   - Crecimiento relativo
   - Distribución geográfica

3. [Análisis de festivos y eventos](#3-análisis-de-festivos-y-eventos)
   - Festivos nacionales
   - Eventos climáticos
   - Conclusiones

4. [Análisis de productos](#4-análisis-de-productos)
   - Ventas por categoría
   - Top / Bottom productos
   - Estacionalidad en productos
   - Productos en declive
   - Productos sin ventas recientes
   - Variación de precios entre tiendas

5. [Matriz de correlación](#5-matriz-de-correlación)

In [ ]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
import plotly.colors as pc
import folium
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
import warnings
warnings.filterwarnings("ignore")


from statsmodels.tsa.seasonal import seasonal_decompose
from plotly.subplots import make_subplots


In [ ]:
try:
    from google.colab import drive
    drive.mount('/content/drive')
    DATA_PATH = '/content/drive/MyDrive/data_dsmarket/'
except ImportError:
    DATA_PATH = 'data_dsmarket/'


df = pd.read_feather(DATA_PATH + 'df_preprocessed.feather')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# **1. Análisis temporal**

Patrones de comportamiento a lo largo del tiempo: tendencia global, estacionalidad trimestral, efecto fin de semana y variación por día y mes.

In [ ]:
df

,id,item,category,department,store,store_code,region,d,sales,date,...,yearweek,sell_price,season,pay_period,ingresos,is_holiday,year,week,month,indice_estacional_store_item
0,ACCESORIES_1_008_NYC_1,ACCESORIES_1_008,ACCESORIES,ACCESORIES_1,Greenwich_Village,NYC_1,New York,d_1,12,2011-01-29,...,201105,0.6118,Invierno,otros,7.3416,0,2011,5,1,2.058577
1,ACCESORIES_1_009_NYC_1,ACCESORIES_1_009,ACCESORIES,ACCESORIES_1,Greenwich_Village,NYC_1,New York,d_1,2,2011-01-29,...,201105,2.0748,Invierno,otros,4.1496,0,2011,5,1,0.726826
2,ACCESORIES_1_010_NYC_1,ACCESORIES_1_010,ACCESORIES,ACCESORIES_1,Greenwich_Village,NYC_1,New York,d_1,0,2011-01-29,...,201105,4.2161,Invierno,otros,0.0000,0,2011,5,1,1.775047
3,ACCESORIES_1_012_NYC_1,ACCESORIES_1_012,ACCESORIES,ACCESORIES_1,Greenwich_Village,NYC_1,New York,d_1,0,2011-01-29,...,201105,7.9534,Invierno,otros,0.0000,0,2011,5,1,1.068961
4,ACCESORIES_1_015_NYC_1,ACCESORIES_1_015,ACCESORIES,ACCESORIES_1,Greenwich_Village,NYC_1,New York,d_1,4,2011-01-29,...,201105,0.9310,Invierno,otros,3.7240,0,2011,5,1,1.779883
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
46027952,SUPERMARKET_3_823_PHI_3,SUPERMARKET_3_823,SUPERMARKET,SUPERMARKET_3,Queen_Village,PHI_3,Philadelphia,d_1913,1,2016-04-24,...,201617,3.5760,Primavera,otros,3.5760,0,2016,17,4,0.461028
46027953,SUPERMARKET_3_824_PHI_3,SUPERMARKET_3_824,SUPERMARKET,SUPERMARKET_3,Queen_Village,PHI_3,Philadelphia,d_1913,0,2016-04-24,...,201617,2.9760,Primavera,otros,0.0000,0,2016,17,4,1.747468
46027954,SUPERMARKET_3_825_PHI_3,SUPERMARKET_3_825,SUPERMARKET,SUPERMARKET_3,Queen_Village,PHI_3,Philadelphia,d_1913,0,2016-04-24,...,201617,4.7760,Primavera,otros,0.0000,0,2016,17,4,0.839307
46027955,SUPERMARKET_3_826_PHI_3,SUPERMARKET_3_826,SUPERMARKET,SUPERMARKET_3,Queen_Village,PHI_3,Philadelphia,d_1913,3,2016-04-24,...,201617,1.5360,Primavera,otros,4.6080,0,2016,17,4,1.111897


## Evolución de ingresos diarios

In [ ]:
serie = df.groupby('date')['ingresos'].sum()

fig = px.line(serie.reset_index(), x='date', y='ingresos',
              title='Ingresos diarios 2011–2016',
              labels={'date': 'Fecha', 'ingresos': 'Ingresos'})
fig.show()

La serie de ingresos diarios muestra una tendencia creciente sostenida a lo largo del periodo analizado,
pasando de ~80k en 2011 hasta alcanzar picos de ~200k en 2016. Se aprecian caídas puntuales a cero que corresponden a los días de cierre de las tiendas

In [ ]:
serie = df.groupby('date')['ingresos'].sum()

fig = go.Figure()

# Serie diaria original (con ruido)
fig.add_trace(go.Scatter(x=serie.index, y=serie.values,
                         name='Ingresos diarios', opacity=0.3,
                         line=dict(color='steelblue')))

# Tendencia a corto plazo (28 días)
fig.add_trace(go.Scatter(x=serie.index, y=serie.rolling(28).mean(),
                         name='Media móvil 28 días',
                         line=dict(color='tomato', width=2)))

# Tendencia a largo plazo (90 días)
fig.add_trace(go.Scatter(x=serie.index, y=serie.rolling(90).mean(),
                         name='Media móvil 90 días',
                         line=dict(color='darkgreen', width=2)))

fig.update_layout(title='Tendencia de ingresos con rolling window')
fig.show()

La de 28 días reacciona más rápido a los cambios porque promedia menos días, así que los picos puntuales todavía la afectan algo.
La de 90 días es mucho más inerte, necesita que algo ocurra de forma sostenida durante 3 meses para que se note, por eso es tan suave.

Las caídas bruscas a cero que se ven en la serie diaria (las líneas que bajan hasta 0) no se propagan a las medias móviles, lo que confirma que son eventos puntuales (cierres, festivos) y no un problema estructural de las ventas.
La tendencia creciente es robusta: se ve tanto en la media de 28 como en la de 90, lo que significa que no es un artefacto del ruido diario sino un crecimiento real sostenido en el tiempo.


## Ingresos por trimestre

 Q1 -> enero, febrero, marzo

 Q2 -> abril, mayo, junio

 Q3 -> julio, agosto, septiembre

 Q4 -> octubre, noviembre, diciembre

In [ ]:
nombres_q = {1: 'Q1', 2: 'Q2', 3: 'Q3', 4: 'Q4'}

ingresos_diarios = (df[df['year'] != 2016]
                    .groupby('date')['ingresos'].sum().reset_index())
ingresos_diarios['year'] = pd.to_datetime(ingresos_diarios['date']).dt.year
ingresos_diarios['quarter'] = pd.to_datetime(ingresos_diarios['date']).dt.quarter

pivot = (ingresos_diarios.groupby(['year', 'quarter'])['ingresos']
                         .mean().reset_index()
                         .pivot(index='year', columns='quarter', values='ingresos'))
pivot.columns = [nombres_q[q] for q in pivot.columns]

fig = px.imshow(pivot,
                title='Ingresos medios diarios por año y trimestre',
                labels=dict(x='Trimestre', y='Año', color='Ingresos medios'),
                color_continuous_scale='Oranges',
                aspect='auto')
fig.show()

El heatmap permite separar dos efectos que en un gráfico de barras se pueden llegar a confundir: la tendencia creciente se lee horizontalmente (filas más oscuras conforme avanzamos en años) y la estacionalidad trimestral se lee verticalmente (la columna Q3 tiende a ser ligeramente más oscura dentro de cada año).

* La tendencia de crecimiento de la cadena es el efecto dominante, pasando de ~80k en 2011 a ~140k en 2015.

* El efecto trimestral existe pero es moderado — Q3 muestra ingresos algo superiores de forma consistente entre años, coherente con el mayor consumo de verano, mientras que Q1 parece ser el trimestre más flojo.

## Efecto fin de semana

In [ ]:
df['is_weekend'] = df['weekday'].isin(['Saturday', 'Sunday']).astype(int)

weekend_comp = (df.groupby(['date', 'is_weekend'])['ingresos']
                  .sum().reset_index()
                  .groupby('is_weekend')['ingresos'].mean().reset_index())

weekend_comp['tipo'] = weekend_comp['is_weekend'].map({0: 'Entre semana', 1: 'Fin de semana'})


fig = px.bar(weekend_comp, x='tipo', y='ingresos',
             title='Ingresos medios diarios: fin de semana vs entre semana',
             labels={'ingresos': 'Ingresos medios diarios', 'tipo': ''})
fig.update_layout(showlegend=False)
fig.show()

Los ingresos medios diarios en fin de semana son aproximadamente un 30% superiores a los días
entre semana, confirmando el efecto fin de semana como un patrón estructural en las ventas de DSMarket.

Desde el punto de vista comercial, los días entre semana representan una oportunidad de acción:
promociones o descuentos puntuales en esos días podrían redistribuir parte de la demanda del
fin de semana y aumentar los ingresos en los días más flojos, optimizando además la capacidad
operativa de las tiendas.

In [ ]:
orden_dias = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
nombres_dias = ['Lunes', 'Martes', 'Miércoles', 'Jueves', 'Viernes', 'Sábado', 'Domingo']
nombres_mes = ['Ene','Feb','Mar','Abr','May','Jun','Jul','Ago','Sep','Oct','Nov','Dic']

pivot = (df[df['year'] != 2016]
           .groupby(['weekday', 'month'])['ingresos']
           .mean().reset_index()
           .pivot(index='weekday', columns='month', values='ingresos'))

pivot = pivot.reindex(orden_dias)
pivot.index = nombres_dias
pivot.columns = nombres_mes

fig = px.imshow(pivot,
                labels=dict(x='Mes', y='Día de la semana', color='Ingresos medios'),
                title='Ingresos medios por día de la semana y mes del año',
                color_continuous_scale='Oranges',
                aspect='auto')
fig.show()

El heatmap confirma el efecto fin de semana de forma visual: sábado y domingo destacan claramente como los días de mayor ingreso en todos los meses del año. Entre semana los ingresos son más homogéneos, aunque lunes y viernes muestran ingresos ligeramente superiores al resto — lunes por las compras de inicio de semana y viernes por la proximidad al fin de semana.

Martes y miércoles son consistentemente los días más flojos, lo que representa una oportunidad comercial directa: promociones o descuentos puntuales en esos días podrían redistribuir parte de la demanda del fin de semana y aumentar los ingresos en los días de menor actividad.

El hueco blanco del jueves de noviembre corresponde a Thanksgiving, festivo en el que las tiendas cierran o reducen drásticamente su actividad.

## Efecto período de cobro

In [ ]:
pay_comp = (df.groupby(['date', 'pay_period'])['ingresos']
              .sum().reset_index()
              .groupby('pay_period')['ingresos'].mean().reset_index())

orden = ['inicio_mes', 'mitad_mes', 'otros']
pay_comp['pay_period'] = pd.Categorical(pay_comp['pay_period'], categories=orden, ordered=True)
pay_comp = pay_comp.sort_values('pay_period')

fig = px.bar(pay_comp, x='pay_period', y='ingresos',
             title='Ingresos medios diarios por período de pago',
             labels={'pay_period': 'Período', 'ingresos': 'Ingresos medios diarios'})
fig.show()

Se observa una ligera variación en los ingresos según el período de pago: tanto a principio como
a mitad de mes los ingresos son algo superiores al resto de días, coherente con los dos períodos
de cobro de nóminas habituales en EEUU. El efecto es algo más pronunciado a inicio de mes.
Sin embargo, las diferencias son moderadas (~5-7%), lo que indica que el efecto existe pero no
es determinante en el comportamiento de compra de esta cadena de supermercados.

# **2. Análisis por tienda**

Evolución de ingresos por tienda, crecimiento relativo y distribución geográfica de los puntos de venta.

## Tendencia de ingresos

In [ ]:
ventas_tienda = df.groupby(['store_code', 'date'])['ingresos'].sum().reset_index()

fig = go.Figure()
for tienda in sorted(ventas_tienda['store_code'].unique()):
    serie = ventas_tienda[ventas_tienda['store_code'] == tienda].set_index('date')['ingresos']
    rolling = serie.rolling(28).mean()
    fig.add_trace(go.Scatter(x=rolling.index, y=rolling.values, name=tienda))

fig.update_layout(title='Tendencia de ingresos por tienda (rolling 28 días)')
fig.show()

El análisis individual de cada tienda revela que todas muestran tendencia alcista sostenida a lo largo del periodo, aunque con comportamientos muy distintos entre ellas.

**NYC_3** destaca claramente como la tienda líder de la cadena, con ingresos que casi doblan a las demás (~25k en 2016). Sus bajadas coinciden con las épocas navideñas (diciembre-enero), un patrón estacional esperado y consistente en todos los años del histórico.

Se detectan anomalías relevantes en varias tiendas:


- **NYC_2**: crecimiento moderado entre 2011 y 2015, seguido de un salto brusco en mayo de 2015 que eleva los ingresos de ~10k a ~16k en pocas semanas. Posible ampliación, reforma o cambio operativo en esa fecha.

- **PHI_1**: caída brusca a finales de octubre de 2012, seguida de una recuperación inmediata y muy pronunciada. El patrón sugiere un posible cierre temporal de la tienda por obras o reforma.

- **PHI_2**: salto brusco a mediados de 2012, doblando sus ingresos en pocos meses. Posible ampliación, reforma o cambio en el surtido de productos de la tienda.

- **PHI_3**: presenta un patrón estacional recurrente con caídas pronunciadas en el período marzo-mayo de cada año (visible en 2012, 2013, 2014 y 2015). La caída de mediados de 2013 es especialmente intensa (-31%). Podría reflejar un comportamiento de consumo estacional local o un problema operativo que se repite en esa franja temporal.

- **BOS_2**: doble caída consecutiva en 2013 sin recuperación completa entre ambas, seguida de un desplome brusco a principios de 2014 (~15%). A partir de ahí inicia la recuperación más fuerte de todas las tiendas de Boston, alcanzando máximos de ~17k en 2016. El comportamiento 2013-2014 es atípico y sin causa climática identificada.

**BOS_1** y **NYC_4** son las tiendas más estables y predecibles, sin anomalías destacables,
lo que las convierte en buenas referencias de comportamiento base para el modelo.

## Patrón semanal


In [ ]:
ventas_dia = df.groupby(['store_code', 'weekday'])['ingresos'].sum().reset_index()
ventas_dia = ventas_dia.pivot(index='weekday', columns='store_code', values='ingresos')
ventas_dia = ventas_dia.reindex(orden_dias)
ventas_dia.index = nombres_dias

# Normalización con fórmula MinMax (0-1) por tienda para que podamos compararlas
ventas_norm = (ventas_dia - ventas_dia.min()) / (ventas_dia.max() - ventas_dia.min())

fig = px.imshow(ventas_norm, color_continuous_scale='Oranges',
                title='Patrón semanal por tienda',
                labels=dict(x='Tienda', y='Día', color='Ingresos norm.'))
fig.show()

Tras normalizar los ingresos por tienda (escala 0–1), el patrón semanal es consistente en toda la cadena: sábado y domingo son los días de mayor actividad en todas las tiendas.

Dentro de los días entre semana, lunes y viernes destacan como los de mayor actividad — patrón ya observado en el análisis temporal. El heatmap normalizado revela que el peso relativo de cada uno varía según la tienda: Boston y varias tiendas de Nueva York muestran un pico más pronunciado el lunes, Philadelphia lo concentra en el viernes, y el resto presenta ambos días con intensidad similar. Esto sugiere diferencias en los hábitos de compra entre ciudades e incluso entre tiendas de la misma ciudad.

## Distribución de ventas diarias

In [ ]:
df_box = df.groupby(['store_code', 'date'])['sales'].sum().reset_index()

fig = px.box(df_box,
             x='store_code', y='sales', color='store_code',
             hover_data=['date'],
             title='Distribución de ventas diarias por tienda',
             labels={'store_code':'Tienda', 'sales':'Ventas diarias'})
fig.update_layout(showlegend=False)
fig.show()

NYC_3 presenta la mediana más alta y mayor variabilidad diaria, mientras que NYC_4 es la tienda más pequeña y con ventas más consistentes.

Los puntos en 0 corresponden al **cierre por Navidad (25 de diciembre)**, confirmado en todas las tiendas.

En el extremo superior se detectan dos outliers coordinados que merecen atención: las **tres tiendas de Philadelphia** alcanzan sus **máximos el mismo día** (9 de abril de 2016) y las **tres tiendas de Boston** comparten uno de sus picos más altos (15 de junio de 2015). El hecho de que el efecto sea simultáneo en todas las tiendas de cada ciudad apunta a un factor externo común — posiblemente un evento local o una promoción coordinada — y no a un comportamiento individual de cada tienda.

## Crecimiento relativo

In [ ]:
crecimientos = []
for tienda in sorted(ventas_tienda['store_code'].unique()):
    serie = ventas_tienda[ventas_tienda['store_code'] == tienda].set_index('date')['ingresos']
    rolling = serie.rolling(28).mean().dropna()
    crecimiento = (rolling.iloc[-1] - rolling.iloc[0]) / rolling.iloc[0] * 100
    crecimientos.append({'tienda': tienda, 'crecimiento': crecimiento})

df_crec = pd.DataFrame(crecimientos).sort_values('crecimiento', ascending=True)

fig = px.bar(df_crec, x='crecimiento', y='tienda', orientation='h',
             title='Crecimiento relativo por tienda 2011-2016 (%)',
             labels={'crecimiento': 'Crecimiento (%)', 'tienda': 'Tienda'})
fig.show()

In [ ]:
# Crecimiento relativo de cada tienda: cuánto ha crecido en % desde 2011 hasta 2016
for tienda in sorted(ventas_tienda['store_code'].unique()):
    serie = ventas_tienda[ventas_tienda['store_code'] == tienda].set_index('date')['ingresos']
    rolling = serie.rolling(28).mean().dropna()
    crecimiento = (rolling.iloc[-1] - rolling.iloc[0]) / rolling.iloc[0] * 100
    print(f"{tienda}: {crecimiento:.1f}%")

BOS_1: 78.2%
BOS_2: 57.2%
BOS_3: 112.2%
NYC_1: 77.4%
NYC_2: 112.0%
NYC_3: 86.7%
NYC_4: 121.1%
PHI_1: 157.1%
PHI_2: 223.1%
PHI_3: 30.6%


In [ ]:
crecimientos_ultimo_año = []
for tienda in sorted(ventas_tienda['store_code'].unique()):
    serie = ventas_tienda[ventas_tienda['store_code'] == tienda].set_index('date')['ingresos']
    rolling = serie.rolling(28).mean().dropna()
    ultimo_año = rolling[rolling.index >= rolling.index.max() - pd.DateOffset(years=1)]
    crecimiento = (ultimo_año.iloc[-1] - ultimo_año.iloc[0]) / ultimo_año.iloc[0] * 100
    crecimientos_ultimo_año.append({'tienda': tienda, 'crecimiento': crecimiento})

df_crec_ult = pd.DataFrame(crecimientos_ultimo_año).sort_values('crecimiento', ascending=True)

fig = px.bar(df_crec_ult, x='crecimiento', y='tienda', orientation='h',
             title='Crecimiento relativo por tienda — último año (%)',
             labels={'crecimiento': 'Crecimiento (%)', 'tienda': 'Tienda'})
fig.show()

In [ ]:
# Crecimiento relativo de cada tienda: cuánto ha crecido en % el último año
for tienda in sorted(ventas_tienda['store_code'].unique()):
    serie = ventas_tienda[ventas_tienda['store_code'] == tienda].set_index('date')['ingresos']
    rolling = serie.rolling(28).mean().dropna()
    ultimo_año = rolling[rolling.index >= rolling.index.max() - pd.DateOffset(years=1)]
    crecimiento = (ultimo_año.iloc[-1] - ultimo_año.iloc[0]) / ultimo_año.iloc[0] * 100
    print(f"{tienda}: {crecimiento:.1f}%")

BOS_1: 3.0%
BOS_2: 4.0%
BOS_3: 2.8%
NYC_1: 4.8%
NYC_2: 78.9%
NYC_3: 6.2%
NYC_4: 14.3%
PHI_1: 14.8%
PHI_2: 37.2%
PHI_3: 25.3%


Comparando el crecimiento histórico (2011–2016) con el del último año se obtienen conclusiones de negocio relevantes:

- **NYC_2** era discreta en el histórico (112%) pero lidera el último año con un 78.9% de crecimiento, consistente con el salto operativo detectado en mayo de 2015. Es la tienda con mayor momentum actual.

- **PHI_2** lidera el histórico (223%) y se mantiene segunda en el último año (37.2%), confirmando que no fue un salto puntual sino un crecimiento sostenido.

- **PHI_3** era la más estancada en el histórico (30.6%) pero sube al tercer puesto en el último año (25.3%), lo que sugiere que podría estar saliendo del patrón estacional negativo detectado anteriormente.

- **Boston** es la región que más preocupa: las tres tiendas ocupan los últimos puestos en el último año (BOS_3 2.8%, BOS_1 3.0%, BOS_2 4.0%), a pesar de que BOS_3 había crecido un 112% en el histórico. La región parece haberse estancado recientemente y merece atención prioritaria.

## Distribución geográfica

Vamos a añadir las coordenadas de cada tienda para poder representarlas en un mapa y ver si hay alguna relación entre la ubicación y las ventas

In [ ]:
df[['store', 'store_code', 'region']].drop_duplicates().sort_values('store_code')

,store,store_code,region
4515,South_End,BOS_1,Boston
5665,Roxbury,BOS_2,Boston
6873,Back_Bay,BOS_3,Boston
0,Greenwich_Village,NYC_1,New York
1204,Harlem,NYC_2,New York
2239,Tribeca,NYC_3,New York
3462,Brooklyn,NYC_4,New York
7972,Midtown_Village,PHI_1,Philadelphia
8888,Yorktown,PHI_2,Philadelphia
9773,Queen_Village,PHI_3,Philadelphia


In [ ]:
coordenadas = {
    'BOS_1': {'store': 'South_End',        'lat': 42.3467, 'lon': -71.0792},
    'BOS_2': {'store': 'Roxbury',          'lat': 42.3188, 'lon': -71.0846},
    'BOS_3': {'store': 'Back_Bay',         'lat': 42.3503, 'lon': -71.0810},
    'NYC_1': {'store': 'Greenwich_Village','lat': 40.7335, 'lon': -74.0027},
    'NYC_2': {'store': 'Harlem',           'lat': 40.8116, 'lon': -73.9465},
    'NYC_3': {'store': 'Tribeca',          'lat': 40.7163, 'lon': -74.0086},
    'NYC_4': {'store': 'Brooklyn',         'lat': 40.6782, 'lon': -73.9442},
    'PHI_1': {'store': 'Midtown_Village',  'lat': 39.9496, 'lon': -75.1609},
    'PHI_2': {'store': 'Yorktown',         'lat': 39.9706, 'lon': -75.1579},
    'PHI_3': {'store': 'Queen_Village',    'lat': 39.9357, 'lon': -75.1527},
}

In [ ]:
ventas_tienda = df.groupby('store_code')['sales'].sum().reset_index()
ventas_tienda['lat'] = ventas_tienda['store_code'].map({k: v['lat'] for k, v in coordenadas.items()})
ventas_tienda['lon'] = ventas_tienda['store_code'].map({k: v['lon'] for k, v in coordenadas.items()})
ventas_tienda['store'] = ventas_tienda['store_code'].map({k: v['store'] for k, v in coordenadas.items()})
ventas_tienda['radio'] = (ventas_tienda['sales'] / ventas_tienda['sales'].max()) * 30000  # meters

color_map = {
    'BOS_1': '#E63946',
    'BOS_2': '#C1121F',
    'BOS_3': '#FF6B6B',
    'NYC_1': '#2196F3',
    'NYC_2': '#0D47A1',
    'NYC_3': '#64B5F6',
    'NYC_4': '#1565C0',
    'PHI_1': '#2E7D32',
    'PHI_2': '#66BB6A',
    'PHI_3': '#1B5E20',
}

offsets = {
    'NYC_1': (0.15,  0.0),
    'NYC_2': (-0.1,  0.15),
    'NYC_3': (0.0,  -0.15),
    'NYC_4': (0.1,   0.1),
    'BOS_1': (0.0,   0.0),
    'BOS_2': (-0.1,  0.1),
    'BOS_3': (0.1,  -0.1),
    'PHI_1': (0.0,   0.0),
    'PHI_2': (-0.1,  0.0),
    'PHI_3': (0.1,   0.0),
}

display(ventas_tienda)

mapa = folium.Map(
    location=[41.5, -73.5],
    zoom_start=7,
    tiles='https://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer/tile/{z}/{y}/{x}',
    attr='Esri World Street Map'
)

for _, row in ventas_tienda.iterrows():
    colorpunto = color_map[row['store_code']]
    dlat, dlon = offsets.get(row['store_code'], (0, 0))
    folium.Circle(
        location=[row['lat'] + dlat, row['lon'] + dlon],
        radius=row['radio'],
        popup=f"{row['store']} ({row['store_code']})<br>Ventas: {row['sales']:,.0f}",
        color=colorpunto,
        fill=True,
        fill_opacity=0.6
    ).add_to(mapa)

mapa

,store_code,sales,lat,lon,store,radio
0,BOS_1,5595292,42.3467,-71.0792,South_End,15003.223044
1,BOS_2,7214384,42.3188,-71.0846,Roxbury,19344.658381
2,BOS_3,6089330,42.3503,-71.0810,Back_Bay,16327.937162
3,NYC_1,7698216,40.7335,-74.0027,Greenwich_Village,20642.006117
4,NYC_2,5685475,40.8116,-73.9465,Harlem,15245.039855
5,NYC_3,11188180,40.7163,-74.0086,Tribeca,30000.000000
6,NYC_4,4103676,40.6782,-73.9442,Brooklyn,11003.602016
7,PHI_1,5149062,39.9496,-75.1609,Midtown_Village,13806.701358
8,PHI_2,6544012,39.9706,-75.1579,Yorktown,17547.122052
9,PHI_3,6427782,39.9357,-75.1527,Queen_Village,17235.462783


El tamaño de cada círculo representa el volumen total de ventas de la tienda. NYC_3 (Tribeca) destaca claramente como la tienda con mayor volumen de toda la cadena. Las tiendas de Nueva York son las más dispersas geográficamente entre sí, mientras que Philadelphia y Boston están más concentradas.

Al hacer click sobre cada círculo se puede consultar el
nombre del barrio y el volumen total de ventas de esa tienda.

# **3. Análisis de festivos y eventos**

Efecto de festivos nacionales y eventos climáticos sobre las ventas. Se identifican los eventos con patrón consistente y repetido a lo largo de los años.

## Festivos nacionales

In [ ]:
eventos_importantes = [
     'Christmas Day', 'NewYear', 'SuperBowl',
    'Easter', 'Thanksgiving', 'Black Friday', 'Cyber Monday', "Valentine's Day",
    'Independence Day', 'Labor Day',
]

for evento in eventos_importantes:
    dias_evento = df[df['event'] == evento]['date'].unique()
    if len(dias_evento) == 0:
        continue

    fig = go.Figure()
    for fecha in sorted(dias_evento):
        ventana = df[
            (df['date'] >= fecha - pd.Timedelta(days=10)) &
            (df['date'] <= fecha + pd.Timedelta(days=10))
        ]
        ventas_ventana = ventana.groupby('date')['sales'].sum().reset_index()
        ventas_ventana['dia_relativo'] = (ventas_ventana['date'] - fecha).dt.days
        fig.add_trace(go.Scatter(
            x=ventas_ventana['dia_relativo'],
            y=ventas_ventana['sales'],
            name=str(fecha)[:4]
        ))

    fig.add_vline(x=0, line_dash='dash', line_color='red')
    fig.update_layout(
        title=f'Ventas ±10 días alrededor de {evento}',
        xaxis_title='Días respecto al evento',
        yaxis_title='Ventas'
    )
    fig.show()

## Eventos climáticos

In [ ]:
evento_fechas = {
    '2012-10-29': 'Hurricane Sandy',
    '2015-01-26': 'Blizzard 2015',
    '2016-01-23': 'Blizzard 2016',
}

fechas = {pd.to_datetime(f): nombre for f, nombre in evento_fechas.items()}

for fecha, nombre in fechas.items():
    ventana = df[(df['date'] >= fecha - pd.Timedelta(days=10)) &
                 (df['date'] <= fecha + pd.Timedelta(days=10))]

    ventas_region = ventana.groupby(['date', 'region'])['sales'].sum().reset_index()
    ventas_region['dia_relativo'] = (ventas_region['date'] - fecha).dt.days

    fig = px.line(ventas_region, x='dia_relativo', y='sales', color='region',
                  title=f'Ventas por región ±10 días — {nombre} ({str(fecha)[:10]})',
                  labels={'dia_relativo': 'Días respecto al evento', 'sales': 'Ventas'})
    fig.add_vline(x=0, line_dash='dash', line_color='red')
    fig.show()

## Conclusiones

### Eventos que incluimos
- **Christmas Day**: caída brusca a 0 el día de navidad todos los años. El efecto más limpio y consistente de todos (tiendas cerradas).
- **NewYear**: caída el día 0 consistente, aunque bastante menos intensa que Christmas.
- **SuperBowl**: gran pico de ventas los días previos al evento, patrón repetido todos los años.
- **Easter**: pico de ventas los días previos, patrón repetido todos los años.
- **Thanksgiving**: caída clara y consistente el día del evento todos los años, con pico de compras anticipadas los días anteriores.

### Eventos que descartamos
- **Black Friday y Cyber Monday**: su efecto queda completamente capturado dentro de la ventana ±10 días de Thanksgiving. La subida posterior a la caída de Thanksgiving responde más a la recuperación natural de la serie que a un efecto propio de estos días.
- **Labor Day**: la curva muestra cierta forma repetida entre años pero 2011 se desmarca claramente del resto, rompiendo la consistencia. Los picos observados en los extremos de la ventana responden al efecto fin de semana, no al evento en sí.
- **Independence Day**: patrón demasiado débil y variable entre años para considerarlo una señal robusta.
- **Valentine's Day**: no se observa patrón consistente entre años.
- **Observed (Christmas, NewYear, Veterans Day)**: caen dentro de la ventana temporal del festivo original y no aportan información adicional.

### Eventos climáticos
- **Hurricane Sandy, Blizzard 2015 y Blizzard 2016**: los tres gráficos muestran el patrón de oscilación semanal habitual sin que se aprecie un efecto claro y consistente atribuible directamente al evento. La única excepción parcial es PHI_1 durante Sandy, cuya caída puntual coincide con las fechas del huracán, aunque la fuerte subida posterior apunta a una reforma o ampliación de la tienda que coincidió en esa misma época.

> El criterio seguido ha sido mantener únicamente los eventos que muestran un patrón consistente y repetido a lo largo de varios años. Los eventos con patrón débil, inconsistente entre años, o cuyo efecto queda capturado por otro evento cercano, no se consideran suficientemente robustos para usarse como features en el modelo.

In [ ]:
def variacion_por_categoria(df, evento, ventana_pre=3):
    fechas_evento = df[df['event'] == evento]['date'].unique()

    # Ingresos totales diarios por categoría
    ventas_diarias = df.groupby(['date', 'category'])['sales'].sum().reset_index()

    # Días con y sin evento
    fechas_con_evento = df[df['event'].notna()]['date'].unique()
    fechas_normales   = ventas_diarias[~ventas_diarias['date'].isin(fechas_con_evento)]['date'].unique()

    # Días previos al evento
    dias_pre = []
    for fecha in fechas_evento:
        mask = (ventas_diarias['date'] >= fecha - pd.Timedelta(days=ventana_pre)) & \
               (ventas_diarias['date'] < fecha)
        dias_pre.append(ventas_diarias[mask])
    df_pre = pd.concat(dias_pre)

    # Días normales
    df_normal = ventas_diarias[ventas_diarias['date'].isin(fechas_normales)]

    media_pre    = df_pre.groupby('category')['sales'].mean()
    media_normal = df_normal.groupby('category')['sales'].mean()

    variacion = ((media_pre - media_normal) / media_normal * 100).round(1)
    variacion.name = evento
    return variacion


eventos_analizar = ['SuperBowl', 'Easter']
tabla = pd.concat([variacion_por_categoria(df, e) for e in eventos_analizar], axis=1)
tabla.columns = eventos_analizar
tabla['categoría'] = tabla.index
tabla = tabla.reset_index(drop=True)
tabla.style.background_gradient(cmap='Greens', subset=['SuperBowl', 'Easter'])\
           .format("{:.1f}%", subset=['SuperBowl', 'Easter'])

,SuperBowl,Easter,categoría
0,19.9%,9.7%,ACCESORIES
1,7.1%,14.8%,HOME_&_GARDEN
2,7.7%,19.7%,SUPERMARKET


Variación porcentual de unidades vendidas en los 3 días previos al evento respecto a un día normal sin evento.

En el SuperBowl, contra lo esperado, ACCESSORIES lidera el crecimiento (+19.9%) por encima de SUPERMARKET (+7.7%), lo que sugiere que el efecto va más allá de la compra de comida y bebida — la gente también compra artículos para ambientar y preparar la celebración en casa.

Easter en cambio sigue un patrón más intuitivo: SUPERMARKET (+19.7%) y HOME_&_GARDEN (+14.8%) lideran, coherente con las compras de comida para celebraciones familiares y la llegada de la primavera.

In [ ]:
# dias 25 de diciembre con ventas distintas de 0
df[(df.date.dt.month == 12) & (df.date.dt.day == 25) & (df.sales != 0)][['date', 'store_code', 'item', 'sales']].sort_values('store_code')

,date,store_code,item,sales
21034071,2013-12-25,BOS_1,SUPERMARKET_3_555,1
42323463,2015-12-25,BOS_1,SUPERMARKET_3_252,1
21036302,2013-12-25,BOS_2,SUPERMARKET_3_080,1
21036709,2013-12-25,BOS_2,SUPERMARKET_3_555,2
21036451,2013-12-25,BOS_2,SUPERMARKET_3_252,3
31301797,2014-12-25,BOS_2,SUPERMARKET_2_125,1
21036734,2013-12-25,BOS_2,SUPERMARKET_3_586,5
31302287,2014-12-25,BOS_2,SUPERMARKET_3_228,1
31302604,2014-12-25,BOS_2,SUPERMARKET_3_555,2
31302442,2014-12-25,BOS_2,SUPERMARKET_3_389,1


Se detectan 60 registros con ventas el 25 de diciembre distribuidos entre varias tiendas y años.
Los volúmenes son mínimos y puntuales, probablemente residuos de datos o productos con reposición
automática. A efectos del modelo, el 25 de diciembre se trata como día de cierre.

# **4. Análisis de categorías**

Evolución temporal, ticket medio y distribución de precios por categoría para entender el peso relativo de cada una y su contribución al negocio.

## Tendencia de ingresos

In [ ]:
ventas_cat = df.groupby(['date', 'category'])['ingresos'].sum().reset_index()

fig = go.Figure()
for cat in sorted(ventas_cat['category'].unique()):
    serie = ventas_cat[ventas_cat['category'] == cat].set_index('date')['ingresos']
    rolling = serie.rolling(28).mean()
    fig.add_trace(go.Scatter(x=rolling.index, y=rolling.values, name=cat))

fig.update_layout(title='Tendencia de ingresos por categoría (rolling 28 días)',
                  xaxis_title='Fecha', yaxis_title='Ingresos')
fig.show()

In [ ]:
for cat in sorted(df['category'].unique()):
    serie = df[df['category'] == cat].groupby('date')['ingresos'].sum()
    rolling = serie.rolling(28).mean().dropna()
    crecimiento = (rolling.iloc[-1] - rolling.iloc[0]) / rolling.iloc[0] * 100
    print(f"{cat}: {crecimiento:.1f}%")

ACCESORIES: 127.1%
HOME_&_GARDEN: 107.5%
SUPERMARKET: 83.0%


- **SUPERMARKET** es la categoría dominante en volumen de ingresos y mantiene un crecimiento sólido y sostenido a lo largo de todo el período (+83%).

- **HOME_&_GARDEN** muestra una estacionalidad muy marcada con subida desde febrero-marzo, pico principal en verano (junio-septiembre) y caída a partir de octubre, patrón coherente con productos de jardinería y decoración exterior. Crece un +107% en el período.

- **ACCESSORIES** es la categoría más pequeña en volumen pero la de mayor crecimiento relativo (+127%), con un salto especialmente pronunciado a principios de 2015 sin causa externa identificada, posiblemente relacionado con cambios operativos internos o de surtido en alguna de las tiendas.

## Ticket medio

In [ ]:
ticket_medio = df.groupby('category').agg(
    ingresos_total=('ingresos', 'sum'),
    ventas_total=('sales', 'sum')
).reset_index()

ticket_medio['ticket_medio'] = ticket_medio['ingresos_total'] / ticket_medio['ventas_total']

fig = px.bar(ticket_medio, x='category', y='ticket_medio', color='category',
             title='Ticket medio por categoría (€ por unidad vendida)',
             labels={'category': 'Categoría', 'ticket_medio': '$ por unidad'})
fig.update_layout(showlegend=False)
fig.show()

**ACCESSORIES** y **HOME_&_GARDEN** presentan un ticket medio por unidad similar, significativamente superior al de **SUPERMARKET**. Sin embargo,
**SUPERMARKET** es con diferencia la categoría que más ingresos genera gracias a un volumen
de unidades vendidas muy superior al resto — vende más barato pero en muchísima mayor
cantidad.

## Distribución de precios

In [ ]:
muestra = df.groupby(['item', 'category'])['sell_price'].mean().reset_index()

fig = px.box(muestra, x='category', y='sell_price', color='category',
             title='Distribución de precios por categoría',
             labels={'category': 'Categoría', 'sell_price': 'Precio (USD)'})
fig.update_layout(showlegend=False)
fig.show()

**SUPERMARKET** es la categoría con precios más concentrados y predecibles (caja estrecha, mayoría de productos entre 2-8$), coherente con productos de consumo diario con precios estables.

**ACCESSORIES** y **HOME_&_GARDEN** presentan mayor dispersión — sus cajas son más anchas y tienen más outliers en la parte alta, lo que refleja un mix de productos con rangos de precio muy distintos dentro de la misma categoría, desde productos baratos hasta artículos de mayor valor. Esto es consistente con el ticket medio más alto observado en ambas categorías.

# **5. Análisis de productos**

Identificación de los productos más y menos rentables, análisis de declive, patrones estacionales y variación de precios entre tiendas.

## Top / Bottom productos

### Top 10 y Bottom 10 globales

In [ ]:
ventas_item = df.groupby('item')['ingresos'].sum().reset_index().sort_values('ingresos', ascending=False)
top10    = ventas_item.head(10)
bottom10 = ventas_item.tail(10).sort_values('ingresos')

fig = make_subplots(rows=1, cols=2,
                    subplot_titles=['Top 10 productos', 'Bottom 10 productos'])
fig.add_trace(go.Bar(x=top10['ingresos'], y=top10['item'], orientation='h',
                     marker_color='steelblue', name='Top 10'), row=1, col=1)
fig.add_trace(go.Bar(x=bottom10['ingresos'], y=bottom10['item'], orientation='h',
                     marker_color='salmon', name='Bottom 10'), row=1, col=2)
fig.update_layout(title='Ingresos totales por producto', height=450, showlegend=False)
fig.show()

Los productos más rentables son mayoritariamente de la categoría SUPERMARKET, con ingresos
totales de hasta 1.5M a lo largo del periodo. Los productos menos rentables pertenecen todos
a ACCESSORIES_2, con ingresos acumulados por debajo de 1.200€ — posibles candidatos a revisión
o descatalogación.

### Top 10 por región

In [ ]:
for region in sorted(df['region'].unique()):
    top_r = (df[df['region'] == region]
               .groupby('item')['ingresos'].sum()
               .nlargest(10).reset_index().sort_values('ingresos'))
    fig = px.bar(top_r, x='ingresos', y='item', orientation='h',
                 title=f'Top 10 productos — {region}',
                 labels={'ingresos':'Ingresos','item':'Producto'})
    fig.show()

Los productos top varían entre ciudades, aunque hay algunos que aparecen en varias regiones
como SUPERMARKET_3_090, SUPERMARKET_3_586 o SUPERMARKET_3_587, lo que sugiere que son
productos de alta demanda transversal.

Boston destaca por tener productos de ACCESSORIES en su top 10 (ACCESORIES_1_354 y
ACCESORIES_1_158), algo que no ocurre en Nueva York ni Philadelphia, donde el top está
dominado casi exclusivamente por SUPERMARKET. Esto sugiere diferencias en el perfil de
consumo entre ciudades que podrían aprovecharse en estrategias de marketing locales.

### Top 10 por categoría

In [ ]:
for cat in sorted(df['category'].unique()):
    top_c = (df[df['category'] == cat]
               .groupby('item')['ingresos'].sum()
               .nlargest(10).reset_index().sort_values('ingresos'))
    fig = px.bar(top_c, x='ingresos', y='item', orientation='h',
                 title=f'Top 10 productos — {cat}',
                 labels={'ingresos':'Ingresos','item':'Producto'})
    fig.show()


Dentro de ACCESSORIES se aprecia una diferencia clara entre subcategorías: los productos de
ACCESORIES_1 alcanzan ingresos de hasta 900k, mientras que los bottom 10 globales eran
todos de ACCESORIES_2, con ingresos por debajo de 1.200€. Esto sugiere perfiles de demanda
muy distintos entre ambas subcategorías.

En HOME_&_GARDEN los ingresos del top 10 son bastante homogéneos, todos entre 400k y 700k,
sin un producto estrella que destaque claramente sobre el resto.

En SUPERMARKET el producto líder (SUPERMARKET_3_586) casi dobla en ingresos al décimo de
la lista, lo que indica una concentración mayor de la demanda en unos pocos productos clave.

### Top 10 por departamento

In [ ]:
for dept in sorted(df['department'].unique()):
    top_d = (df[df['department'] == dept]
               .groupby('item')['ingresos'].sum()
               .nlargest(10).reset_index().sort_values('ingresos'))
    if top_d.empty: continue
    fig = px.bar(top_d, x='ingresos', y='item', orientation='h',
                 title=f'Top 10 productos — {dept}',
                 labels={'ingresos':'Ingresos','item':'Producto'})
    fig.show()

El análisis por departamento revela diferencias muy marcadas entre subcategorías:

* Dentro de **ACCESSORIES**, el departamento _1 supera en hasta 13 veces al _2 en ingresos del producto líder (900k vs 70k), lo que refuerza la hipótesis de que ACCESORIES_2 es una subcategoría con demanda muy débil.

* En **HOME_&_GARDEN** se repite el patrón: _1 triplica a _2.

* En **SUPERMARKET**, el departamento _3 concentra los productos más rentables de toda la cadena.

## Elasticidad precio-demanda

### Elasticidad en productos top

Del análisis de los productos más vendidos por categoría, seleccionamos 4 productos con comportamientos interesantes en términos de relación entre precio y demanda. El fondo de color indica el nivel de precio relativo de cada semana (verde = precio bajo, rojo = precio alto).

In [ ]:
productos_analisis = [
    'SUPERMARKET_3_586',
    'SUPERMARKET_3_090',
    'HOME_&_GARDEN_1_413',
    'ACCESORIES_1_354'
]

for producto in productos_analisis:
    # Ventas diarias
    ventas_prod = df[df['item'] == producto].groupby('date')['sales'].sum().reset_index()

    # Precio semanal medio
    precios_prod = (df[df['item'] == producto]
                    .groupby('yearweek')['sell_price'].mean()
                    .reset_index()
                    .sort_values('yearweek'))

    # Convertir yearweek a fecha
    precios_prod['date'] = pd.to_datetime(precios_prod['yearweek'].astype(str) + '1', format='%Y%W%w')

    # Normalizar precios
    p_min = precios_prod['sell_price'].min()
    p_max = precios_prod['sell_price'].max()
    precios_prod['precio_norm'] = (precios_prod['sell_price'] - p_min) / (p_max - p_min)

    fig = go.Figure()

    for i in range(len(precios_prod) - 1):
        x0 = precios_prod['date'].iloc[i]
        x1 = precios_prod['date'].iloc[i + 1]
        color_val = precios_prod['precio_norm'].iloc[i]
        color = pc.sample_colorscale('RdYlGn_r', [color_val])[0]
        fig.add_vrect(x0=x0, x1=x1, fillcolor=color, opacity=0.3, line_width=0)

    fig.add_trace(go.Scatter(x=ventas_prod['date'], y=ventas_prod['sales'],
                             name='Ventas', line=dict(color='black', width=1.5)))

    fig.update_layout(title=f'Ventas y precio — {producto}',
                      xaxis_title='Fecha', yaxis_title='Unidades vendidas')
    fig.show()

- **SUPERMARKET_3_586**: precio sube progresivamente con los años sin impacto en ventas — producto inelástico de consumo habitual.
- **SUPERMARKET_3_090**: los períodos de precio bajo coinciden exactamente con los picos de ventas más altos — elasticidad precio-demanda clara.
- **HOME_&_GARDEN_1_413**: precio alto inicial con ventas casi nulas; al bajar el precio las ventas se activan y crecen de forma sostenida.
- **ACCESORIES_1_354**: descuento puntual en enero 2015 actúa como activador de demanda latente — las ventas se disparan y se mantienen elevadas incluso cuando el precio vuelve a subir, sugiriendo que el descuento generó una base de clientes nueva y estable.

### Elasticidad en productos en declive

In [ ]:
# Agrupación mensual por producto
df['year_month'] = df['date'].dt.to_period('M')
ventas_mensuales = df.groupby(by = ['item', 'year_month']).sales.sum().reset_index()
ventas_mensuales['year_month'] = ventas_mensuales['year_month'].astype(str)
df

,id,item,category,department,store,store_code,region,d,sales,date,...,season,pay_period,ingresos,is_holiday,year,week,month,indice_estacional_store_item,is_weekend,year_month
0,ACCESORIES_1_008_NYC_1,ACCESORIES_1_008,ACCESORIES,ACCESORIES_1,Greenwich_Village,NYC_1,New York,d_1,12,2011-01-29,...,Invierno,otros,7.3416,0,2011,5,1,2.058577,1,2011-01
1,ACCESORIES_1_009_NYC_1,ACCESORIES_1_009,ACCESORIES,ACCESORIES_1,Greenwich_Village,NYC_1,New York,d_1,2,2011-01-29,...,Invierno,otros,4.1496,0,2011,5,1,0.726826,1,2011-01
2,ACCESORIES_1_010_NYC_1,ACCESORIES_1_010,ACCESORIES,ACCESORIES_1,Greenwich_Village,NYC_1,New York,d_1,0,2011-01-29,...,Invierno,otros,0.0000,0,2011,5,1,1.775047,1,2011-01
3,ACCESORIES_1_012_NYC_1,ACCESORIES_1_012,ACCESORIES,ACCESORIES_1,Greenwich_Village,NYC_1,New York,d_1,0,2011-01-29,...,Invierno,otros,0.0000,0,2011,5,1,1.068961,1,2011-01
4,ACCESORIES_1_015_NYC_1,ACCESORIES_1_015,ACCESORIES,ACCESORIES_1,Greenwich_Village,NYC_1,New York,d_1,4,2011-01-29,...,Invierno,otros,3.7240,0,2011,5,1,1.779883,1,2011-01
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
46027952,SUPERMARKET_3_823_PHI_3,SUPERMARKET_3_823,SUPERMARKET,SUPERMARKET_3,Queen_Village,PHI_3,Philadelphia,d_1913,1,2016-04-24,...,Primavera,otros,3.5760,0,2016,17,4,0.461028,1,2016-04
46027953,SUPERMARKET_3_824_PHI_3,SUPERMARKET_3_824,SUPERMARKET,SUPERMARKET_3,Queen_Village,PHI_3,Philadelphia,d_1913,0,2016-04-24,...,Primavera,otros,0.0000,0,2016,17,4,1.747468,1,2016-04
46027954,SUPERMARKET_3_825_PHI_3,SUPERMARKET_3_825,SUPERMARKET,SUPERMARKET_3,Queen_Village,PHI_3,Philadelphia,d_1913,0,2016-04-24,...,Primavera,otros,0.0000,0,2016,17,4,0.839307,1,2016-04
46027955,SUPERMARKET_3_826_PHI_3,SUPERMARKET_3_826,SUPERMARKET,SUPERMARKET_3,Queen_Village,PHI_3,Philadelphia,d_1913,3,2016-04-24,...,Primavera,otros,4.6080,0,2016,17,4,1.111897,1,2016-04


In [ ]:
def calcular_tendencia_producto(df):
   # esta función calcula la pendiente de la tendencia de las ventas mensuales de un producto usando descomposición estacional y regresión lineal sobre la tendencia
    try:
        if len(df) < 24:
            return np.nan
        resultado = seasonal_decompose(df['sales'], model = 'additive', period=6)
        tendencia = resultado.trend.dropna()    # quito el principio y el final pq se me quedan muchos nulos
        x = np.arange(len(tendencia))
        pendiente = np.polyfit(x, tendencia, 1)[0]
        return pendiente
    except:
        return np.nan


tendencias_productos = []
for item in ventas_mensuales['item'].unique():
    producto = ventas_mensuales[ventas_mensuales['item'] == item].sort_values('year_month')
    pendiente = calcular_tendencia_producto(producto)
    tendencias_productos.append({
        'item': item,
        'pendiente_tendencia': pendiente,
        'ventas_totales': producto['sales'].sum()
    })
df_tendencias = pd.DataFrame(tendencias_productos)
df_tendencias = df_tendencias.dropna()
df_tendencias = df_tendencias.sort_values('pendiente_tendencia')
print("productos con pendiente + negativa:")
print(df_tendencias.head(20))

productos con pendiente + negativa:
                   item  pendiente_tendencia  ventas_totales
2374  SUPERMARKET_3_150           -96.357509          147532
2186  SUPERMARKET_2_360           -78.836104          257119
2779  SUPERMARKET_3_555           -70.333456          491287
2810  SUPERMARKET_3_586           -63.211339          920242
2765  SUPERMARKET_3_541           -57.068168          272624
2859  SUPERMARKET_3_635           -56.388234          282134
3021  SUPERMARKET_3_800           -55.282599           90197
1954  SUPERMARKET_2_128           -55.132040          149465
2314  SUPERMARKET_3_090           -50.233397         1002529
1848  SUPERMARKET_2_021           -43.940724          121679
2888  SUPERMARKET_3_664           -43.906241           68270
2341  SUPERMARKET_3_117           -43.742361           64951
2378  SUPERMARKET_3_154           -43.141725          101583
2670  SUPERMARKET_3_446           -41.866947           96285
2692  SUPERMARKET_3_468           -39.319181     

In [ ]:
productos_declive = df_tendencias.head(10)
fig = make_subplots(rows=5, cols=2, subplot_titles=[f"{prod}" for prod, pend in zip(productos_declive['item'], productos_declive['pendiente_tendencia'])])

for idx, prod in enumerate(productos_declive['item']):
    datos_prod = ventas_mensuales[ventas_mensuales['item'] == prod].sort_values('year_month')
    fig.add_trace(go.Scatter(x=datos_prod['year_month'], y=datos_prod['sales'], mode='lines+markers', name=prod, showlegend=False), row=idx//2+1, col=idx%2+1)

fig.update_layout(height=1200, title_text="Top 10 Productos en Declive")
fig.show()

In [ ]:
productos_declive_vrect = [
    'SUPERMARKET_3_150',
    'SUPERMARKET_3_800',
    'SUPERMARKET_3_555',
    'SUPERMARKET_3_586',
    'SUPERMARKET_2_021',
    'SUPERMARKET_2_128'
]

for producto in productos_declive_vrect:
    ventas_prod = df[df['item'] == producto].groupby('date')['sales'].sum().reset_index()

    precios_prod = (df[df['item'] == producto]
                    .groupby('yearweek')['sell_price'].mean()
                    .reset_index()
                    .sort_values('yearweek')
                    .dropna(subset=['sell_price']))

    precios_prod['date'] = pd.to_datetime(precios_prod['yearweek'].astype(str) + '1', format='%Y%W%w')

    p_min = precios_prod['sell_price'].min()
    p_max = precios_prod['sell_price'].max()

    if p_max == p_min:
        continue

    precios_prod['precio_norm'] = (precios_prod['sell_price'] - p_min) / (p_max - p_min)

    fig = go.Figure()

    for i in range(len(precios_prod) - 1):
        x0 = precios_prod['date'].iloc[i]
        x1 = precios_prod['date'].iloc[i + 1]
        color = pc.sample_colorscale('RdYlGn_r', [precios_prod['precio_norm'].iloc[i]])[0]
        fig.add_vrect(x0=x0, x1=x1, fillcolor=color, opacity=0.3, line_width=0)

    fig.add_trace(go.Scatter(x=ventas_prod['date'], y=ventas_prod['sales'],
                             name='Ventas', line=dict(color='black', width=1.5)))

    fig.update_layout(title=f'Ventas y precio — {producto}',
                      xaxis_title='Fecha', yaxis_title='Unidades vendidas')
    fig.show()


Del análisis de los 10 productos en mayor declive, se identifican 3 patrones distintos interesantes, cada uno representado por un par de productos con comportamiento muy similar — posiblemente variantes o sustitutivos dentro de la misma familia de producto:

- **SUPERMARKET_3_150 / SUPERMARKET_3_800**: declive gradual sostenido coincidente con subida progresiva de precio (verde → rosa). El precio parece ser un factor clave en la pérdida de demanda. En 2015 hay un período prolongado con ventas prácticamente a cero antes de que el precio baje de nuevo a finales de 2016 (fondo verde), momento en que las ventas repuntan — confirmando la elasticidad.

- **SUPERMARKET_3_555 / SUPERMARKET_3_586**: declive más moderado y resistente al precio. A pesar de la subida progresiva de precio, las ventas se mantienen en niveles razonables — productos con demanda más estable e inelástica.

- **SUPERMARKET_2_021 / SUPERMARKET_2_128**: durante el período de precio bajo (verde) las ventas son estables. En la fase intermedia (amarillo) hay incluso un repunte inicial antes de iniciar el declive gradual. Ya en el período de precio alto (rosa) las ventas caen a 0 en varios momentos y el nivel general se reduce significativamente.

## Productos sin ventas recientes

In [ ]:
# Ventana de 12 meses
ultimo_año = df['date'].max() - pd.DateOffset(months=12)
df_reciente = df[df['date'] >= ultimo_año].copy()

# Ventas diarias por producto
ventas_reciente = df_reciente.groupby(['item', 'date'])['sales'].sum().reset_index()
ventas_reciente = ventas_reciente.sort_values(['item', 'date'])

# Días desde hoy para calcular peso (más reciente = más peso)
fecha_max = ventas_reciente['date'].max()
ventas_reciente['dias_desde_hoy'] = (fecha_max - ventas_reciente['date']).dt.days

# Peso exponencial decreciente: e^(-dias/30) — escala mensual
ventas_reciente['peso'] = np.exp(-ventas_reciente['dias_desde_hoy'] / 30)

# Score de inactividad: suma de pesos en días con ventas = 0
ventas_reciente['es_cero'] = (ventas_reciente['sales'] == 0).astype(int)
ventas_reciente['score_inactividad'] = ventas_reciente['peso'] * ventas_reciente['es_cero']

ranking = (ventas_reciente.groupby('item')['score_inactividad']
           .sum()
           .sort_values(ascending=False)
           .reset_index())

print(ranking.head(10))

                  item  score_inactividad
0     ACCESORIES_1_335          30.502629
1    SUPERMARKET_3_441          30.441094
2    SUPERMARKET_3_444          30.436967
3  HOME_&_GARDEN_2_202          29.567418
4  HOME_&_GARDEN_2_456          29.373187
5  HOME_&_GARDEN_2_502          29.349121
6  HOME_&_GARDEN_2_210          29.341473
7     ACCESORIES_2_110          29.231081
8    SUPERMARKET_3_210          28.922808
9    SUPERMARKET_1_126          28.545572


In [ ]:
top10_inactividad = ranking.head(10)['item'].tolist()

fig = make_subplots(rows=5, cols=2,
                    subplot_titles=top10_inactividad)

for idx, prod in enumerate(top10_inactividad):
    datos = ventas_mensuales[ventas_mensuales['item'] == prod].sort_values('year_month')
    fig.add_trace(go.Scatter(x=datos['year_month'], y=datos['sales'],
                             mode='lines+markers', name=prod, showlegend=False),
                  row=idx//2+1, col=idx%2+1)

fig.update_layout(height=1200, title_text='Histórico de ventas — top 10 productos con mayor score de inactividad')
fig.show()

El análisis de los 10 productos con mayor score de inactividad ponderada (últimos 12 meses) revela tres perfiles distintos:

- **Descatalogados**: productos que vendían en un período concreto y cayeron a cero de forma definitiva (ACCESORIES_1_335). Probablemente retirados del catálogo.

- **Intermitentes o estacionales**: productos con períodos a cero intercalados con ventas regulares a lo largo de todo el histórico (HOME_&_GARDEN_2_202, HOME_&_GARDEN_2_456, HOME_&_GARDEN_2_502, HOME_&_GARDEN_2_210, SUPERMARKET_3_441, SUPERMARKET_3_444). No están inactivos, simplemente tienen demanda intermitente o estacional — el modelo deberá tratarlos de forma diferenciada.

- **En declive gradual**: productos con tendencia descendente sostenida que han llegado a cero recientemente (SUPERMARKET_3_210, ACCESORIES_2_110, SUPERMARKET_1_126). Candidatos a revisión o descatalogación por el equipo comercial.

## Productos con mas diferencia de precios entre tiendas

In [ ]:
precios = df.groupby('item')['sell_price'].describe() # tenemos los 3049 productos, vamos a quedarnos con los q presentasen mayor varianza en los precios
precios['rango'] = precios['max'] - precios['min']
precios.sort_values('rango', ascending=False).head(10)

,count,mean,std,min,25%,50%,75%,max,rango
item,,,,,,,,,
HOME_&_GARDEN_2_406,19109.0,16.073263,6.074483,4.0750,15.575000,15.5750,15.5875,134.149994,130.074994
HOME_&_GARDEN_2_466,18472.0,9.425621,5.512242,7.4625,8.712500,8.7125,8.7125,65.775002,58.312501
HOME_&_GARDEN_2_178,18451.0,8.743999,1.586798,3.7500,8.712500,8.7125,8.7125,55.450001,51.700001
HOME_&_GARDEN_2_250,19102.0,8.797407,1.376150,4.2000,8.700000,8.7000,8.7125,42.724998,38.524999
HOME_&_GARDEN_2_514,19109.0,23.466715,1.151297,1.2500,22.424999,23.7125,23.7125,26.212500,24.962500
ACCESORIES_1_342,13908.0,26.289127,1.230137,2.6600,25.230101,27.1054,27.1054,27.105400,24.445400
HOME_&_GARDEN_1_469,18934.0,22.467644,0.881044,1.2500,22.462500,22.4625,22.4625,25.575001,24.325001
ACCESORIES_1_257,18430.0,22.346489,0.516967,10.5868,22.290800,22.2908,22.2908,34.313999,23.727200
HOME_&_GARDEN_1_342,7937.0,22.392048,0.915705,1.2500,22.400000,22.4625,22.4625,22.462500,21.212500


In [ ]:
precios.sort_values(by = 'std', ascending = False).head(10)

,count,mean,std,min,25%,50%,75%,max,rango
item,,,,,,,,,
HOME_&_GARDEN_2_406,19109.0,16.073263,6.074483,4.075000,15.575000,15.575000,15.587500,134.149994,130.074994
HOME_&_GARDEN_2_466,18472.0,9.425621,5.512242,7.462500,8.712500,8.712500,8.712500,65.775002,58.312501
HOME_&_GARDEN_2_292,19046.0,15.810708,3.148933,12.462500,12.462500,18.712500,18.712500,19.962500,7.500000
ACCESORIES_1_060,13873.0,39.486069,2.066095,34.580002,39.846802,39.860100,41.203400,41.203400,6.623398
ACCESORIES_1_361,13943.0,39.487301,2.063049,35.883400,39.846802,39.860100,41.203400,41.203400,5.320000
ACCESORIES_1_225,13943.0,39.494667,2.060246,35.883400,39.846802,39.860100,41.203400,41.203400,5.320000
HOME_&_GARDEN_1_259,19130.0,12.143476,2.010479,7.487500,9.350000,13.712500,13.712500,14.400000,6.912499
ACCESORIES_1_311,17926.0,20.972946,1.994851,17.290001,19.497801,19.497801,23.647400,23.913401,6.623400
HOME_&_GARDEN_2_446,19074.0,31.822943,1.853013,18.737499,31.150000,32.462502,32.462502,35.000000,16.262501


* Rango de precios (max - min): mide la diferencia entre la tienda más cara y la más barata para cada producto. Un rango alto indica que el mismo producto se vende a precios muy diferentes según la ubicación.
* Desviación estándar: mide cuánto varían los precios en general. Mientras que el rango solo captura los extremos, la desviación estándar considera todos los precios y detecta si la variación es generalizada o si es solo una tienda atípica.

Se utilizan dos métricas complementarias: el rango captura los extremos de precio
y la desviación estándar detecta si la variación es generalizada o puntual.

El análisis por rango y desviación estándar identifica los productos con mayor variabilidad
de precio. HOME_&_GARDEN_2_406 destaca como el caso más extremo, con un rango de 130€ entre
el precio mínimo y máximo registrado — el mismo producto se vende desde 4€ hasta 134€ según
la tienda y el período. Esta variabilidad puede reflejar promociones puntuales, diferencias
de pricing entre ciudades o errores en los datos que conviene investigar.

In [ ]:
top_rango = precios.sort_values('rango', ascending=False).head(10).index.tolist()
df_top = df[df['item'].isin(top_rango)]

fig = px.box(df_top, x='item', y='sell_price',
             title='Top 19 productos con mayor variación de precio',
             labels={'item': 'Producto', 'sell_price': 'Precio (USD)'})
fig.show()

Los productos con mayor variabilidad de precio muestran outliers puntuales muy alejados del precio habitual — especialmente HOME_&_GARDEN_2_406 con un máximo de ~135 USD frente a un precio típico de ~4 USD. Esta variabilidad puede reflejar promociones puntuales, diferencias de pricing entre ciudades o errores en los datos que conviene investigar
antes del modelado.

# **6. Matriz de correlación**

Análisis de correlación entre las variables numéricas del dataset.

In [ ]:
num_cols = [c for c in ['sales', 'sell_price', 'ingresos', 'is_holiday', 'week', 'month', 'year'] if c in df.columns]
corr = df[num_cols].corr().round(2)
fig = px.imshow(corr, text_auto=True, color_continuous_scale='RdBu', color_continuous_midpoint=0, title='Correlación entre variables numéricas', width=700, height=600)
fig.show()

La correlación entre `week` y `month` es casi perfecta (0.97), lo que indica redundancia — en el modelo de predicción bastará con usar una de las dos.

`sales` e `ingresos` tienen una correlación alta (0.7) como es esperable.

`sell_price` y `sales` muestran una correlación negativa (-0.15), consistente con el efecto de elasticidad precio-demanda observado en el análisis de productos: a mayor precio, menor volumen de ventas.

`is_holiday` no muestra correlación lineal significativa con ninguna variable, aunque su efecto sí es visible en ventanas temporales concretas como se analizó en la sección de festivos.